client/trino-client/src/main/java/io/trino/client/auth/kerberos/DelegatedConstrainedContextProvider.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.client.auth.kerberos;
+
+import io.airlift.units.Duration;
+import org.ietf.jgss.GSSContext;
+import org.ietf.jgss.GSSCredential;
+import org.ietf.jgss.GSSException;
+import org.ietf.jgss.GSSManager;
+
+import javax.security.auth.login.LoginException;
+
+import static java.util.Objects.requireNonNull;
+import static org.ietf.jgss.GSSContext.INDEFINITE_LIFETIME;
+import static org.ietf.jgss.GSSName.NT_HOSTBASED_SERVICE;
+
+public class DelegatedConstrainedContextProvider
+        implements GSSContextProvider
+{
+    protected static final GSSManager GSS_MANAGER = GSSManager.getInstance();
+
+    private final GSSCredential gssCredential;
+
+    public DelegatedConstrainedContextProvider(GSSCredential gssCredential)
+    {
+        this.gssCredential = requireNonNull(gssCredential, "gssCredential is null");
+    }
+
+    @Override
+    public GSSContext getContext(String servicePrincipal, Duration minCredentialLifetime)
+            throws GSSException, LoginException
+    {
+        GSSContext gssContext = GSS_MANAGER.createContext(
+                GSS_MANAGER.createName(servicePrincipal, NT_HOSTBASED_SERVICE),
+                SPNEGO_OID,
+                gssCredential,
+                INDEFINITE_LIFETIME);
+
+        gssContext.requestMutualAuth(true);
+        gssContext.requestConf(true);
+        gssContext.requestInteg(true);
+        gssContext.requestCredDeleg(true);
+        return gssContext;
```

Comment:
Can we extract this method to `GSSContextProvider` where we have a method like `GSSContext createGSSContext(String servicePrincipal, GSSCredential clientCredential)` and `GSSContextProvider` could be an abstract class which share the `GSS_MANAGER`. WDYT ?

Reply:
I was thinking if we could extract the way of how we create the GSSContext based on the GSSCredential then we could reuse them at other ContextProvider right ? Like either we could invoke them directly or invoke them via doAs

Reply:
I think, I got it, refactored